### PR TITLE
Add ability to prune ledger

### LINKED ISSUE #5082: TdS ledger rollback procedure
#5041 and #5057 move us towards being able to perform a cluster rollback.  What remains is:
1. https://github.com/solana-labs/solana/blob/1c966aac25c88cc3cc876ea011b9dc7e025b5802/ledger-tool/src/main.rs#L194
2. Document the manual rollback procedure as a markdown file in https://github.com/solana-labs/tour-de-sol
3. Test the procedure and make sure the cluster can actually restart after a rollback

### PR DESCRIPTION
**Pull request by sagar-solana**
#### Problem

Ledger tool's `prune` does not actually prune the ledger

#### Summary of Changes

Add functions to prune the ledger. Add preliminary tests.

Remaining: 
- [x] Check that cluster reboots from a pruned ledger
- [x] Rollback procedure documentation (https://github.com/solana-labs/tour-de-sol/pull/3)

Fixes #5082

### CODE REVIEW

File: core/src/blocktree.rs [hunk 1]
```diff
@@ -193,6 +193,46 @@ impl Blocktree {
...
+            max_index = index;
+        }
+        for i in 0..=max_index {
+            column.delete((delete_slot, i))?;
+        }
+        Ok(())
+    }
+
+    fn delete_slot(&self, slot: u64) -> Result<()> {
+        match self.meta(slot) {
```
Comment (rob-solana) [inline core/src/blocktree.rs]: I don't think you need to predicate the delete_indexed_column() on whether or not meta is present for the slot...  delete with extreme prejudice, do not assume that meta is necessarily present, or truthy

File: core/src/blocktree.rs [hunk 2]
```diff
@@ -193,6 +193,46 @@ impl Blocktree {
...
+            Ok(meta) => match meta {
+                None => Ok(()),
+                Some(_) => {
+                    self.meta_cf.delete(slot)?;
+                    self.delete_indexed_column(&self.data_cf, slot)?;
+                    self.delete_indexed_column(&self.erasure_meta_cf, slot)?;
```
Comment (rob-solana) [inline core/src/blocktree.rs]: why would we bail on io errors in this prune?  better to bail on a consistency check?

**Comment by mvines** [comment]
> Rollback procedure documentation (Separate PR?)

Yep, that info goes here: https://github.com/solana-labs/tour-de-sol/blob/master/README-internal.md#ledger-rollback-procedure

File: core/src/blocktree.rs [hunk 3]
```diff
@@ -512,6 +552,11 @@ impl Blocktree {
         self.data_cf.get_bytes((slot, index))
     }
 
+    /// Dangerous. Use with care.
```
Comment (mvines) [inline core/src/blocktree.rs]: nit: describe the danger?

File: core/src/blocktree.rs [hunk 4]
```diff
@@ -921,6 +966,41 @@ impl Blocktree {
...
+        meta.next_slots.clear();
+        self.put_meta_bytes(
+            target_slot,
+            &bincode::serialize(&meta).expect("couldn't get meta bytes"),
+        )
+        .expect("unable to update meta");
+
+        self.slot_meta_iterator(target_slot + 1)
+            .expect("unable to iterate over meta")
+            .for_each(|(slot, _)| self.delete_slot(slot).unwrap());
```
Comment (rob-solana) [inline core/src/blocktree.rs]: ok, errors thrown to this point.  how about an expect() instead?

Comment (rob-solana) [reply core/src/blocktree.rs]: this code is going to be in heavy use in the near term, and we don't know what errors are gonna pop out, we need to log the hell out of it in order to best support the community

File: core/src/blocktree.rs [hunk 5]
```diff
@@ -921,6 +966,41 @@ impl Blocktree {
...
+            self.put_meta_bytes(
+                slot,
+                &bincode::serialize(&meta).expect("couldn't update meta"),
+            )
+            .expect("couldn't update meta");
+            if slot > target_slot {
+                break;
+            }
+        }
+    }
```
Comment (rob-solana) [inline core/src/blocktree.rs]: consistency check at the end of this function?

Comment (rob-solana) [reply core/src/blocktree.rs]: or maybe another function that looks like fsck() that a tool can call

File: core/src/blocktree.rs (hunk 1 above)
Comment (sagar-solana) [reply core/src/blocktree.rs]: Yeah you're right. I think Im going to rework this PR to basically iter each column as far as possible and wipe it.

File: core/src/blocktree.rs (hunk 2 above)
Comment (sagar-solana) [reply core/src/blocktree.rs]: I will log all errors and attempt continue with the deletion.

File: core/src/blocktree.rs (hunk 4 above)
Comment (sagar-solana) [reply core/src/blocktree.rs]: Yeah agreed, I think I rushed into marking this as ready. Was just glad it even worked lol

File: core/src/blocktree.rs (hunk 5 above)
Comment (sagar-solana) [reply core/src/blocktree.rs]: filed this as a separate issue https://github.com/solana-labs/solana/issues/5156

**Pull request merged**